### PR TITLE
Run SDK simulations with tgrade

### LINKED ISSUE #70: Run simulations in CI
Add support and run one of them

* State export test is **blocked** by #50 
* TestFullAppSimulation ~~is currently blocked by https://github.com/confio/tgrade-contracts/issues/485~~ most modules actually work with utgd.

### PR DESCRIPTION
**Pull request by alpe**
Resolves #70 

Includes 
* testing contract upgrade to v1.0.0

### CODE REVIEW

File: x/poe/simulation/operations.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+			TxGen:           simappparams.MakeTestEncodingConfig().TxConfig,
+			Cdc:             nil,
+			Msg:             msg,
+			MsgType:         msg.Type(),
+			Context:         ctx,
+			SimAccount:      simAccount,
+			AccountKeeper:   ak,
+			Bankkeeper:      bk,
+			ModuleName:      types.ModuleName,
+			CoinsSpentInMsg: spendable,
```
Comment (alpe) [inline x/poe/simulation/operations.go:200]: this should be empty, unless you pay fees

File: x/poe/simulation/operations.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+
+		validators, _, err := k.ValsetContract(ctx).ListValidators(ctx, nil)
+		if len(validators) == 0 || err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgDelegate, "numbers of validator equal zero"), nil, nil
+		}
+
+		simAccount, _ := simtypes.RandomAcc(r, accs)
```
Comment (alpe) [inline x/poe/simulation/operations.go:218]: Unlike PoS only operators can delegate. So instead you would select a random validator and delegate from their validator operator address. The code would same as L160-176 and then delegate the spendable amount

File: x/poe/simulation/operations.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+			App:           app,
+			TxGen:         simappparams.MakeTestEncodingConfig().TxConfig,
+			Cdc:           nil,
+			Msg:           msg,
+			MsgType:       msg.Type(),
+			Context:       ctx,
+			SimAccount:    simAccount,
+			AccountKeeper: ak,
+			ModuleName:    types.ModuleName,
+		}
```
Comment (alpe) [inline x/poe/simulation/operations.go:259]: Does it work without CoinsSpentInMsg?

File: x/poe/simulation/operations.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+		msg := types.NewMsgDelegate(simAccount.Address, selfDelegation, sdk.NewCoin(denom, sdk.ZeroInt()))
+
+		txCtx := simulation.OperationInput{
+			App:           app,
+			TxGen:         simappparams.MakeTestEncodingConfig().TxConfig,
+			Cdc:           nil,
+			Msg:           msg,
+			MsgType:       msg.Type(),
+			Context:       ctx,
+			SimAccount:    simAccount,
```
Comment (alpe) [inline x/poe/simulation/operations.go:256]: this would also be the validator operator account

File: x/poe/simulation/operations.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+		validators, _, err := k.ValsetContract(ctx).ListValidators(ctx, nil)
+		if len(validators) == 0 || err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "numbers of validator equal zero"), nil, nil
+		}
+
+		val := validators[rand.Intn(len(validators))]
+		valAddr, err := sdk.AccAddressFromBech32(val.OperatorAddress) // plain string to not run into bech32 prefix issues with valoper
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "validator address is not valid"), nil, err
+		}
```
Comment (alpe) [inline x/poe/simulation/operations.go:279]: L270-279 look very familiar, now. Would it make sense to extract this to a helper function?

File: x/poe/simulation/operations.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+		delegated, err := k.EngagementContract(ctx).QueryDelegated(ctx, valAddr)
```
Comment (alpe) [inline x/poe/simulation/operations.go:280]: Good query

File: x/poe/simulation/operations.go [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+		unbondAmt, err := simtypes.RandPositiveInt(r, totalBond)
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "invalid unbond amount"), nil, err
+		}
+
+		if unbondAmt.IsZero() {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "unbond amount is zero"), nil, nil
+		}
+
+		delAddr, err := sdk.AccAddressFromBech32(delegated.Delegated) // plain string to not run into bech32 prefix issues with valoper
```
Comment (alpe) [inline x/poe/simulation/operations.go:299]: you can skip this and use delAddr .

File: x/poe/simulation/operations.go [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+		if err != nil {
+			return simtypes.NoOpMsg(types.ModuleName, types.TypeMsgUndelegate, "delegated address is not valid"), nil, err
+		}
+		msg := types.NewMsgUndelegate(
+			delAddr, sdk.NewCoin(k.GetBondDenom(ctx), unbondAmt),
```
Comment (alpe) [inline x/poe/simulation/operations.go:304]: nice

File: x/poe/simulation/operations.go [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+		)
+
+		// need to retrieve the simulation account associated with delegation to retrieve PrivKey
+		var simAccount simtypes.Account
+
+		for _, simAcc := range accs {
```
Comment (alpe) [inline x/poe/simulation/operations.go:310]: you can use `simtypes.FindAccount` instead

File: x/poe/simulation/operations.go [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+			TxGen:           simappparams.MakeTestEncodingConfig().TxConfig,
+			Cdc:             nil,
+			Msg:             msg,
+			MsgType:         msg.Type(),
+			Context:         ctx,
+			SimAccount:      simAccount,
+			AccountKeeper:   ak,
+			Bankkeeper:      bk,
+			ModuleName:      types.ModuleName,
+			CoinsSpentInMsg: spendable,
```
Comment (alpe) [inline x/poe/simulation/operations.go:334]: this should be empty unless you paid fees

**Review by alpe — COMMENTED** [review]
Nice progress. There are some minor issues with adopting for PoE

File: .circleci/config.yml [hunk 11]
```diff
@@ -111,6 +111,19 @@ jobs:
       - store_artifacts:
           path: /go/src/github.com/confio/tgrade/testnet/node3.out
 
+  simulations:
+    executor: golang
+    parallelism: 1
+    resource_class: large
```
Comment (pinosu) [inline .circleci/config.yml:117]: 👍

File: app/params/weights.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package params
+
+// Default simulation operation weights for messages and gov proposals
+const (
+	DefaultWeightMsgCreateValidator int = 100
+	DefaultWeightMsgUpdateValidator int = 5
+	DefaultWeightMsgDelegate        int = 100
+	DefaultWeightMsgUndelegate      int = 100
```
Comment (pinosu) [inline app/params/weights.go:8]: nice

File: x/globalfee/module.go [hunk 13]
```diff
@@ -132,3 +134,26 @@ func (a AppModule) EndBlock(context sdk.Context, block abci.RequestEndBlock) []a
 func (am AppModule) ConsensusVersion() uint64 {
 	return 1
 }
+
+// GenerateGenesisState genesis state for simulations only. Set to empty global fee
```
Comment (pinosu) [inline x/globalfee/module.go:138]: 👍

File: x/twasm/simulation/decoder.go [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/kv"
+)
+
+// NewDecodeStore returns a decoder function closure that unmarshals the KVPair's
+// Value to the corresponding twasm type.
+// Not fully implemented but falls back into default output
+func NewDecodeStore() func(kvA kv.Pair, kvB kv.Pair) string {
+	return func(kvA, kvB kv.Pair) string {
+		switch {
```
Comment (pinosu) [inline x/twasm/simulation/decoder.go:19]: nice!

**Review by pinosu — APPROVED** [review]
LGTM. really good work 🥇

**Comment by daniellarita** [comment]
moved this into the sprint board!

**Pull request merged**